sklearn/cluster/tests/test_k_medoids.py

```diff
@@ -0,0 +1,164 @@
+"""Testing for K-Medoids"""
+import numpy as np
+
+from sklearn.exceptions import NotFittedError
+from sklearn.utils.testing import assert_equal, assert_raises
+from sklearn.utils.testing import assert_true
+from sklearn.utils.testing import assert_greater
+from sklearn.utils.testing import assert_array_equal
+from sklearn.utils.testing import raises
+
+from sklearn.metrics.pairwise import PAIRWISE_DISTANCE_FUNCTIONS
+from sklearn.metrics.pairwise import euclidean_distances
+from sklearn.datasets import load_iris
+
+from sklearn.cluster import KMedoids, KMeans
+
+
+rng = np.random.RandomState(0)
+X = rng.rand(100, 5)
+
+
+@raises(ValueError)
+def test_kmedoids_fit_fails_n_clusters_is_zero():
+    # n_clusters is 0
+    KMedoids(n_clusters=0).fit(X)
+
+
+@raises(ValueError)
+def test_kmedoids_fit_fails_n_clusters_is_none():
+    # n_clusters is None
+    KMedoids(n_clusters=None).fit(X)
+
+
+@raises(ValueError)
+def test_kmedoids_fit_fails_bad_clustering_method():
+    # Bad clustering
+    KMedoids(n_clusters=5, clustering_method='foo').fit(X)
+
+
+@raises(ValueError)
+def test_kmedoids_fit_fails_init_is_none():
+    KMedoids(init=None).fit(X)
+
+
+def test_kmedoids_fit_succeeds():
+    model = KMedoids().fit(X)
+
+    assert_true(model is not None)
+
+    model = KMedoids(n_clusters=5).fit(X)
+
+    assert_true(model is not None)
+
+    model = KMedoids(n_clusters=5, clustering_method='pam').fit(X)
+
+    assert_true(model is not None)
+
+
+def test_kmedoids_fit_fails_too_few_samples_vs_clusters():
+    model = KMedoids(n_clusters=8)
+
+    Xsmall = rng.rand(5, 2)
+
+    # Trying to fit 3 samples to 8 clusters -> Wrong!
+    assert_raises(ValueError, model.fit, Xsmall)
+
+
+def test_kmedoids_fails_predict_before_fit():
+    model = KMedoids()
+
+    assert_raises(NotFittedError, model.predict, X)
+
+
+def test_kmedoids_fails_transform_before_fit():
+    model = KMedoids()
+
+    assert_raises(NotFittedError, model.transform, X)
+
+
+def test_kmedoids_fit_naive_with_all_pairwise_distance_functions():
+    for distance_metric in PAIRWISE_DISTANCE_FUNCTIONS.values():
+        if distance_metric is None:
+            continue
+
+        model = KMedoids(n_clusters=3, distance_metric=distance_metric)
+
```

Comment:
That's the toughest one :) Hope it's ok the way I did it, however, I don't quite 'feel' this style, hence there might be still something to fix